controller/forwarding/internal.go

```diff
@@ -0,0 +1,182 @@
+	}
+
+	return c, nil
+}
+
+// Validate returns an error if any of the internal controller's field is not valid.
+func (c *InternalController) Validate() error {
+	if c.logger == nil {
+		return core.ErrNilPointer.Wrap("logger is required for the internal controller")
+	}
+	if c.BaseController == nil {
+		return core.ErrNilPointer.Wrap("base controller is required for the internal controller")
+	}
+	if c.handler == nil {
+		return core.ErrNilPointer.Wrap("handler is required for the internal controller")
+	}
+
+	return nil
+}
+
+// HandlePacket implements types.ForwardingController.
+func (c *InternalController) HandlePacket(
+	ctx context.Context,
+	packet *types.ForwardingPacket,
+) error {
+	c.logger.Debug("Handling internal packet")
+
+	if packet == nil {
+		return errorsmod.Wrap(core.ErrNilPointer, "internal controller received nil packet")
+	}
+
+	attr, err := c.ExtractAttributes(packet.Forwarding)
+	if err != nil {
+		return errorsmod.Wrap(err, "error extracting internal forwarding attributes")
+	}
+
+	err = c.ValidateForwarding(ctx, packet.TransferAttributes, attr)
+	if err != nil {
+		return errorsmod.Wrap(err, "error validating internal forwarding")
+	}
+
+	err = c.executeForwarding(
+		ctx,
+		packet.TransferAttributes,
+		attr,
+		packet.Forwarding.PassthroughPayload,
+	)
+	if err != nil {
+		return errorsmod.Wrap(err, "Hyperlane controller execution error")
+	}
+
+	return nil
+}
+
+// ExtractAttributes returns the internal attributes from the forwarding or an error.
+func (c *InternalController) ExtractAttributes(
+	forwarding *core.Forwarding,
+) (*forwardingtypes.InternalAttributes, error) {
+	attr, err := forwarding.CachedAttributes()
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "error extracting cached attributes")
+	}
+
+	intAttr, ok := attr.(*forwardingtypes.InternalAttributes)
+	if !ok {
+		return nil, sdkerrors.ErrInvalidType.Wrapf(
+			"expected %T, got %T",
+			&forwardingtypes.InternalAttributes{},
+			attr,
+		)
+	}
+
+	return intAttr, nil
+}
+
+// ValidateForwarding checks whether the forwarding attributes are valid or not.
+func (c *InternalController) ValidateForwarding(
+	ctx context.Context,
+	transferAttr *types.TransferAttributes,
+	intAttr *forwardingtypes.InternalAttributes,
+) error {
+	if err := transferAttr.Validate(); err != nil {
+		return errorsmod.Wrap(err, "invalid transfer attributes")
+	}
+
+	if err := intAttr.Validate(); err != nil {
+		return errorsmod.Wrap(err, "invalid transfer attributes")
+	}
+
+	return nil
+}
+
+// executeForwarding initiates an internal transfer.
+func (c *InternalController) executeForwarding(
+	ctx context.Context,
+	transferAttr *types.TransferAttributes,
+	intAttr *forwardingtypes.InternalAttributes,
+	_ []byte,
+) error {
+	_, err := c.handler.Send(ctx, &banktypes.MsgSend{
+		FromAddress: core.ModuleAddress.String(),
+		ToAddress:   intAttr.Recipient,
+		Amount: sdk.Coins{
+			sdk.NewCoin(transferAttr.DestinationDenom(), transferAttr.DestinationAmount()),
+		},
+	})
```

Comment:
in the adapter implementation we are using the keeper methods directly (`SendCoinsFromModuleToModule`) -- is there a particular reason that you decided to use the bank msg server here instead?